### ISSUE TITLE
Visualizing the predictions

### ISSUE THREAD

**Issue by melisekm** [mareksubocz/TrackNet#4]
Hey, thanks for good work, I would like to ask about the `predict.py`. I was using it before, but it seemed that the ball position was little bit behind the image as if it was drawing the bbox, or circle on the first frame of the prediction instead of all three (but for each frame). 

Now you did some changes (here https://github.com/mareksubocz/TrackNet/commit/9621f58e453fe572f0e6a1c32ab51a35dc1bdf49)  and if I understand it correctly, for each frame you are drawing three frames? so the resulting video would have three times the frames? I am trying to do something here, but now it feels like there is one prediction per three frames, what I am doing wrong? Thanks for help

```python
def get_ball_position(img):
    ret, thresh = cv.threshold(img, 0.5, 1, 0)
    thresh = cv.convertScaleAbs(thresh)

    contours, hierarchy = cv.findContours(
        thresh, cv.RETR_EXTERNAL, cv.CHAIN_APPROX_NONE
    )

    if len(contours) != 0:
        c = max(contours, key=cv.contourArea)
        x, y, w, h = cv.boundingRect(c)
        return x, y, w, h  # coco

...
cap = cv.VideoCapture(f"{VIDEOS_FILE_PATH}/{video_name}.mp4")
prev_frame_1 = None
prev_frame_2 = None
img_array = []
frame_arr = []
while cap.isOpened():
    success, frame = cap.read()
    if not success:
        break
    frame_arr.append(frame)
    frame_torch = torch.tensor(frame).permute(2, 0, 1).float().to(device) / 255

    if prev_frame_1 is not None and prev_frame_2 is not None:
        frames = torch.cat(
            [prev_frame_2, prev_frame_1, frame_torch], dim=0
        ).unsqueeze(0)
        pred = model(frames)
        pred = pred[0, :, :, :].detach().cpu().numpy()
        if len(frame_arr) == 3:
            for i in range(pred.shape[0]):
                pred_cut = pred[i, :, :]
                frame_cut = frame_arr[i]
                pos = get_ball_position(pred_cut)
                if pos:
                    pos = format_ball(*pos)
                    anotate(frame_cut, pos[0])
                img_array.append(frame_cut)
            frame_arr = []

    prev_frame_2 = prev_frame_1
    prev_frame_1 = frame_torch
... # save to mp4
```

EDIT: I tried to use the old version, but instead of working with `pred = pred[0, 0, :, :].detach().cpu().numpy()` I simply pick `pred = pred[0, 2, :, :].detach().cpu().numpy()` which corresponds with the frame that is currently being read, Ill stick to that until I figure out why the for loop is not working.

Also do you have an idea how to get more than one ball from prediction? 
or get confidence of prediction, similiar to how other object detection models work. I got the threshold set to 0.5, but whats the exact value.

**Comment by mareksubocz** [comment]
Hi, I'm glad this repo gets some interest :) To be honest, ```predict.py``` is not yet finished, but I actually intended to finish it this week. Till now i just used it to quickly preview how a trained network performs on an unseen-before video. 

By default, TrackNet **takes** 3 consecutive frames ( number controlled by parameter ```sequence_length``` ) and **outputs** three consecutive heatmaps as well. If you want to output just one frame, use ```one_output_frame``` option during training, the middle frame will be outputed. 

Quick tip if you didn't realize, ```predict.py``` displays two video windows, but sometimes the second one is hidden behind the first one and you need to move it manually. Watching them both might answer some of your questions.

As the model is heatmap-based, the brightness of the output pixel might be interpreted as the confidence value. Threshold equal to 0.5 means all the pixels with brightness in the range (0.5, 1] are considered white and those in range [0, 0.5] are considered black. After that, ```predict.py``` finds the biggest blob of white in the heatmap using cv.findContours  and displays it as a detected object.

It would certainly be possible to detect more balls, but you would have to come up with a condition on what is a ball and what is not, for example you could display all contours bigger than some value instead of choosing the biggest one. In this case it might also be helpful to have a dataset with more than one ball, but you would have to modify ```dataset.py``` to draw it. 

In my personal use-case I always only have one ball at a time, but PRs are super welcome, I'm also happy to help along the way if you get stuck :)

Let me know if something is still unclear.

**Comment by mareksubocz** [comment]
3481287aec369cfdfb44b3625d86bf3ffdf8886a does this commit help?

**Comment by melisekm** [comment]
I was not using the `cv.imshow` as I was saving the annotated frames into list and then writing to video later in the script. 
But yes the update commit has helped. Thank you.

Regarding the other information, thanks for explaining, It is certainly more clear now.  We can close this issue.
